src/onnx/include/migraphx/onnx/onnx_parser.hpp

```
@@ -97,6 +97,7 @@ struct onnx_parser
     std::unordered_map<std::string, instruction_ref> instructions;
     program prog                                   = program();
     shape::dynamic_dimension default_dyn_dim_value = {1, 1};
+    size_t default_dim_value                       = 0;
```

This should be set to 1. The value `0` is not a valid value.

---

https://github.com/ROCm/AMDMIGraphX/pull/3770#discussion_r1927456684

---

This should be named `batch_dimension` to make it clearer its setting the batch dimension. Also the `--default-dim` flag should be added to the driver to enable the old behavior, since the `--batch` flag is used to set a default dimension.

---

It isn't so much about the internal variable names, but the realization that we might need to preserve the default behavior. But else we should raise the exception as it is. Thus, either an environment variable or a command line option should explicitly turn on the old behavior.

---

Internal variable names are still important for clarity(as I thought this did something else), but I am also talking about the driver flags. I dont think you understand what the `--batch` flag does. It does 2 things:

- Sets the default dim for parameterized dims in the onnx model(similar to `--default-dyn-dim`)
- Adjust the rate on the perf report

The problem you are trying to fix is related to the first item because it will inadvertently set the window dimensions. We can have the `--batch` set only for the first dimension when its a paramterized dim, and when its not, we just ignore it so it can still adjust the perf report. The only way for use to know in the parser that we want to set the first dimension is by having a variable such as `batch_dimension` in the onnx parser.

However, we have used the `--batch` flag to set the default dims in the past as its easier than using `--default-dyn-dim` flag for that. So to make this complete we should also add a `--default-dim` flag which will set the `default_dyn_dim_value` variable in the parser. This should be a simple change to the driver.

---

Please see changes in program.cpp. I was trying to address both the issues that you have mentioned, not just one.

---

> The problem you are trying to fix is related to the first item because it will inadvertently set the window dimensions. We can have the `--batch` set only for the first dimension when its a paramterized dim, and when its not, we just ignore it so it can still adjust the perf report. The only way for use to know in the parser that we want to set the first dimension is by having a variable such as `batch_dimension` in the onnx parser.

@lakhinderwalia this is very close to what my thoughts were at the end of the call. Although my preference would be to not just ignore the command line argument when the batch is not a paramterized dim, I'd prefer to have it error out so the user is very aware that the batch option is not applicable to the onnx file they are running. Ignoring it to "hack" perf reports will only enable QA and others to run things like batch-sweeps on onnx files that do not support it.

---

> I'd prefer to have it error out so the user is very aware that the batch option is not applicable to the onnx file they are running. Ignoring it to "hack" perf reports will only enable QA and others to run things like batch-sweeps on onnx files that do not support it.

The `--batch` flag is used on onxx files that have the batch already set to update the perf report. What do you suggest we do to update the perf report with the batch size? It would be too much of a breaking change to change this.

---

> > @lakhinderwalia this is very close to what my thoughts were at the end of the call. Although my preference would be to not just ignore the command line argument when the batch is not a paramterized dim, I'd prefer to have it error out so the user is very aware that the batch option is not applicable to the onnx file they are running. Ignoring it to "hack" perf reports will only enable QA and others to run things like batch-sweeps on onnx files that do not support it.

This PR now _additionally_ prints a `raw rate` besides the previously printed rate. So all the prior expectations are not broken in any existing test-scripts.

---

> Ignoring it to "hack" perf reports will only enable QA and others to run things like batch-sweeps on onnx files that do not support it.

Also I dont think there is an easy way to check for an error to avoid this case without preventing valid cases. The current error checking doesnt check if the batch is unused, it only is an error if there is a parametrized dimension, which means we will still have an issue with batch-sweeps on onxx files that dont have a dynamic batch.

Also, throwing an error when it finds a parameterized dim thats not the first dimension will prevent valid uses cases beyond changing the perf report. For example, if a model takes two parameters and the first parameter has batch and the second parameter has a parametrized dim thats not the first, then setting the batch and using the default dim on the second parameter will throw an error when there is no user error. Furthermore `parse_type` can't see all the parameters to check for such errors, and refactoring to handle all these scenarios would just make the code too complicated that it is not worth it.

Either way, if QA does a batch-sweep on a model that doesnt support it, we will still get a bug report, but instead of an error about batch it will be a slower rate. Since we are adding a "Raw Rate" we can easily see that the batch was never changing.

---

hmm I see, I don't really like the idea of --batch argument having 2 different meanings/functions (ie. **sometimes** changing input shape for compile, and **always** modifying the QPS in perf report).  Ideally a perf report should just say "Given input shape `[a, b, c, d]`, the model takes `x` s to execute" and that's it. Anything more should really be for the user/tester to interpret (ie. User/tester can decide that since the shape is `[a, b, c, d]`, the batch size is `a` and so the "Rate" is `a/x`). 

But i guess its too late for that now since the current calculation is expected in many workflows, I think its ok to ignore and move on in the static batch dim for now. But I do still think that --batch should only try to alter index 0 of an input shape, and if there are more dynamic dims then just ask the user to specify shapes using --input-dim

---

We are adding an additional statistic for a `Raw rate`, so that hopefully helps about the perf numbers. And we don't throw any exceptions here, and keep the current workflow.

The main reason for the exception is, a serious issue: avoiding an input dimension of size `1px`x`1px`! Also, we have had another case, where the `encoder_sequence_length` was being default-set to `1` -- Luckily this PR caught it.